model_signing/signature/signing.py

```diff
@@ -0,0 +1,209 @@
+                        encoding=crypto_serialization.Encoding.PEM,
+                        format=crypto_serialization.PublicFormat.SubjectPublicKeyInfo,
+                    ),
+                    key_details=common_pb.PublicKeyDetails.PKIX_ECDSA_P256_SHA_256,
+                ),
+            ),
+            dsse_envelope=env,
+        )
+        return bdl
+
+
+class SigstoreSigner(Signer):
+    """Provides a Signer that uses sigstore for signing."""
+
+    CLIENT_ID = "sigstore"
+
+    def __init__(self, disable_ambient: bool = True, id_provider: str = None):
+        token = self.__get_identity_token(disable_ambient, id_provider)
+        if not token:
+            raise ValueError("No identity token supplied or detected!")
+        log.info(f"Signing identity provider: {token.expected_certificate_subject}")
+        log.info(f"Signing identity: {token.identity}")
+        self._signer = sign.Signer(
+            identity_token=token,
+            signing_ctx=sign.SigningContext.production(),
+        )
+
+    @staticmethod
+    def __convert_stmnt(stmnt: statement.Statement) -> dsse.Statement:
+        subjects = stmnt.pb.subject
+        sigstore_subjects = []
+        for s in subjects:
+            sigstore_subjects.append(
+                dsse._Subject(
+                    name=s.name,
+                    digest={"sha256": s.digest["sha256"]},
+                )
+            )
+        return dsse._StatementBuilder(
+            predicate_type=stmnt.pb.predicate_type,
+            predicate=stmnt.pb.predicate,
+            subjects=sigstore_subjects,
+        ).build()
+
+    @staticmethod
+    def __get_identity_token(
+        disable_ambient: bool = True,
+        id_provider: Optional[str] = None,
+    ) -> Optional[oidc.IdentityToken]:
+        token: oidc.IdentityToken
+        if not disable_ambient:
+            return oidc.detect_credential()
+
+        issuer = oidc.Issuer(id_provider) if id_provider else oidc.Issuer.production()
+        token = issuer.identity_token(
+            client_id=SigstoreSigner.CLIENT_ID, force_oob=True
+        )
+        return token
+
+    def sign(self, stmnt: statement.Statement) -> bundle_pb.Bundle:
+        return self._signer.sign_dsse(self.__convert_stmnt(stmnt))._inner
+
+
+class KeySigner(Signer):
+    """Provides a Signer using an elliptic curve private key for signing."""
+
+    def __init__(self, private_key_path: str, password: Optional[str] = None):
+        self._private_key: crypto_ec.EllipticCurvePrivateKey
+        with open(private_key_path, "rb") as fd:
+            serialized_key = fd.read()
+        self._private_key = crypto_serialization.load_pem_private_key(
+            serialized_key, password=password
+        )
+
+    def sign(self, stmnt: statement.Statement) -> bundle_pb.Bundle:
+        pae = utils.pae(stmnt.pb)
+        sig = self._private_key.sign(pae, crypto_ec.ECDSA(SHA256()))
+        env = intoto_pb.Envelope(
```

Comment:
I think we'll want to sign an abstract Manifest class (introduced in https://github.com/sigstore/model-transparency/pull/190) rather than intoto statement. Will let @mihaimaruseac comment